drivers/timer/stm32_lptim_timer.c

```diff
@@ -0,0 +1,273 @@
+
+	/* configure the LPTIM1 counter */
+	LL_LPTIM_SetClockSource(LPTIM1, LL_LPTIM_CLK_SOURCE_INTERNAL);
+	/* configure the LPTIM1 prescaler is 2 */
+	LL_LPTIM_SetPrescaler(LPTIM1, LL_LPTIM_PRESCALER_DIV2);
+	LL_LPTIM_SetPolarity(LPTIM1, LL_LPTIM_OUTPUT_POLARITY_REGULAR);
+	LL_LPTIM_SetUpdateMode(LPTIM1, LL_LPTIM_UPDATE_MODE_IMMEDIATE);
+	LL_LPTIM_SetCounterMode(LPTIM1, LL_LPTIM_COUNTER_MODE_INTERNAL);
+	LL_LPTIM_DisableTimeout(LPTIM1);
+	/*LL_LPTIM_EnableResetAfterRead(LPTIM1);*/
+	LL_LPTIM_TrigSw(LPTIM1);
+	LL_LPTIM_SetInput1Src(LPTIM1, LL_LPTIM_INPUT1_SRC_GPIO);
+	LL_LPTIM_SetInput2Src(LPTIM1, LL_LPTIM_INPUT2_SRC_GPIO);
+
+	/* LPTIM1 interrupt set-up before enabling  */
+	/* no Compare match Interrupt */
+	LL_LPTIM_DisableIT_CMPM(LPTIM1);
+	LL_LPTIM_ClearFLAG_CMPM(LPTIM1);
+
+	/* Autoreload match Interrupt */
+	LL_LPTIM_EnableIT_ARRM(LPTIM1);
+	LL_LPTIM_ClearFLAG_ARRM(LPTIM1);
+
+	/* Enable the LPTIM1 counter */
+	LL_LPTIM_Enable(LPTIM1);
+#ifdef CONFIG_TICKLESS_KERNEL
+	/* Set the Autoreload value once the timer is enabled */
+	LL_LPTIM_SetAutoReload(LPTIM1, LPTIM_TIMEBASE);
+
+	/* do not use compare yet, set the compare to period */
+	LL_LPTIM_SetCompare(LPTIM1, 0);
+
+	/* Start the LPTIM counter in continuous mode */
+	counter1 = 0;
+	last_cnt = 0;
+	LL_LPTIM_StartCounter(LPTIM1, LL_LPTIM_OPERATING_MODE_CONTINUOUS);
+
+#endif /* CONFIG_TICKLESS_KERNEL */
+
+#ifdef CONFIG_DEBUG
+	/* stop LPTIM1 during DEBUG */
+	LL_DBGMCU_APB1_GRP1_FreezePeriph(LL_DBGMCU_APB1_GRP1_LPTIM1_STOP);
+#endif
+	return 0;
+}
+
+void z_clock_set_timeout(s32_t ticks, bool idle)
+{
+	/* LPTIM1 AutoReload value aligned on Kernel ticks */
+	u32_t next_arr = 0;
+
+	ARG_UNUSED(idle);
+#ifdef CONFIG_TICKLESS_KERNEL
+
+	if (ticks == K_FOREVER) {
+		/* disable LPTIM */
+		LL_APB1_GRP1_ForceReset(LL_APB1_GRP1_PERIPH_LPTIM1);
+		LL_APB1_GRP1_DisableClock(LL_APB1_GRP1_PERIPH_LPTIM1);
+		return;
+	}
+
+	/* passing ticks==1 means "announce the next tick",
+	 * ticks value of zero (or even negative) is legal and
+	 * treated identically: it simply indicates the kernel would like the
+	 * next tick announcement as soon as possible.
+	 */
+	if (ticks <= (s32_t)1) {
+		ticks = 1;
+	} else {
+		ticks = (ticks - 1);
+	}
+	/* evaluate the remaining nb of cycle to next tick */
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	/* read current counter value */
+	u32_t cnt_value = LL_LPTIM_GetCounter(LPTIM1);
+
+	/* adjust the next ARR match value to align on Ticks
+	 * from the current counter value to first next Tick
+	 */
+	next_arr = ((cnt_value / COUNT_PER_TICK) + 1)
+		* COUNT_PER_TICK - cnt_value;
+	/* add ms from the expected nb of Ticks */
+	next_arr = next_arr + __ticks_to_ms(ticks) * COUNT_PER_MS;
+
+	/* if too short (ie 8 counter cycles) to trig timer,
+	 * then add one more tick (or fire IRQ)
+	 */
+	if (next_arr < 8) {
+		next_arr = next_arr + COUNT_PER_TICK;
+	}
+
+	/* maximise to TIMEBASE */
+	if (next_arr > LPTIM_TIMEBASE) {
+		next_arr = LPTIM_TIMEBASE;
+	}
```

Comment:
Will counter automatically restart from 0 ? 
Otherwise in case next_arr < cnt_value, you need to also program to LPTIM_TIMEBASE ?

Reply:
yes, LPTIM is reset when ARR and CNT register values matches